pkg/serviceaccount/externaljwt/metrics/metrics.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"crypto/sha256"
+	"errors"
+	"sync"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const (
+	namespace = "apiserver"
+	subsystem = "externaljwt"
+)
+
+var (
+	failedKeyFetch = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "fetch_jwks_error_total",
+			Help:           "Total failed attempts at syncing supported JWKs",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"grpc_status_code"},
+	)
+
+	totalKeyFetch = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "fetch_jwks_request_total",
+			Help:           "Total attempts at syncing supported JWKs",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"status"},
+	)
+
+	lastKeyFetchTimeStamp = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "jwks_last_fetch_time",
+			Help:           "Timestamp of the last successful or failed automatic reload of encryption configuration split by apiserver identity.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		nil,
+	)
+
+	tokenGenFailure = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "sign_jwt_error_total",
+			Help:           "Total failed attempts at signing JWT",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"grpc_status_code"},
+	)
+
+	tokenGenReqTotal = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "sign_jwt_request_total",
+			Help:           "Total attempts at signing JWT",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"status"},
+	)
```

Comment:
cc @ahmedtd as well for feedback on metrics details

Reply:
For metrics purely related to the connection to the plugin, I would use the `WithChainUnaryInterceptor` approach that KMS v2 uses to wire in the metrics at the gRPC level:

https://github.com/kubernetes/kubernetes/blob/175a5b9c4690c63b7d41c3c295402269780b3a27/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service.go#L76

That gives you the total, latency, along with any label dimensions such as method/code.

For any higher order metrics like 'the last time the public keys changed' (just making up an example), we would need more specific wiring.

Reply:
> For metrics purely related to the connection to the plugin, I would use the `WithChainUnaryInterceptor` approach that KMS v2 uses to wire in the metrics at the gRPC level:

hrmm... if FetchKeys succeeds at the GRPC level and fails semantic validation, it seems ugly to have a second failure metric that captures the semantic failure... does KMS have two metrics you have to monitor for failure?

Reply:
I think we can probably make do with one histogramvec recording outbound gRPC calls, tagged by the gRPC method and gRPC status code.  Operation counts can be trivially derived from the histogram, there's no need for separate counters.

We can collect the metrics using a gRPC unary interceptor --- see grpc.WithChainUnaryInterceptor as the client option.

It would also be helpful to get a gauge vector that tells us the key IDs currently in the key cache.  See https://gist.github.com/ahmedtd/ce30522b437bda9b063f67f8750036bc for how we did it internally.

Reply:
now I have:

- `fetch_keys_request_total` with "code" label to record requests to plugin.
- `sign_request_total` with "code" label to record requests to plugin.
- `outbound_request` with "method" and "code" label and hooked onto rpc conn via ChainInterceptor to record Latencies and request counts to external-jwt-signer.

removed separate success/total counts.